src/main/flight/gps_rescue.c

```diff
@@ -458,34 +520,58 @@ static void performSanityChecks()
         rescueState.failure = RESCUE_LOWSATS;
     }
+
+    DEBUG_SET(DEBUG_RTH, 3, (secondsStalled * 100 + secondsLowSats * 10 + secondsFlyingAway)); //Failure can change with no new GPS Data
 }
 
 static void sensorUpdate()
 {
     rescueState.sensor.currentAltitudeCm = getEstimatedAltitudeCm();
     rescueState.sensor.healthy = gpsIsHealthy();
+    static timeUs_t previousDataTimeUs = 0;
+    static float prevDistanceToHomeCm = 0.0f;
 
-    // Calculate altitude velocity
-    static uint32_t previousTimeUs;
-    static int32_t previousAltitudeCm;
-
-    const uint32_t currentTimeUs = micros();
-    const float dTime = currentTimeUs - previousTimeUs;
+    if (rescueState.phase == RESCUE_LANDING) {
+        // do this at sensor update rate, not the much slower GPS rate
+        rescueState.sensor.accMagnitude = (float) sqrtf(sq(acc.accADC[Z]) + sq(acc.accADC[X]) + sq(acc.accADC[Y])) * acc.dev.acc_1G_rec;
+    }
 
-    if (newGPSData) { // Calculate velocity at lowest common denominator
-        rescueState.sensor.distanceToHomeM = GPS_distanceToHome;
+    if (newGPSData) {
+        rescueState.sensor.distanceToHomeCm = GPS_distanceToHomeCm;
+        rescueState.sensor.distanceToHomeM = rescueState.sensor.distanceToHomeCm / 100.0f;
+        rescueState.sensor.groundSpeedCmS = gpsSol.groundSpeed; // cm/s
         rescueState.sensor.directionToHome = GPS_directionToHome;
         rescueState.sensor.numSat = gpsSol.numSat;
-        rescueState.sensor.groundSpeed = gpsSol.groundSpeed;
+        rescueState.sensor.errorAngle = (attitude.values.yaw - rescueState.sensor.directionToHome) * 0.1f;
+        // both attitude and direction are in degrees * 10, errorAngle is degrees
+        if (rescueState.sensor.errorAngle <= -180) {
+            rescueState.sensor.errorAngle += 360;
+        } else if (rescueState.sensor.errorAngle > 180) {
+            rescueState.sensor.errorAngle -= 360;
+        }
 
-        rescueState.sensor.zVelocity = (rescueState.sensor.currentAltitudeCm - previousAltitudeCm) * 1000000.0f / dTime;
-        rescueState.sensor.zVelocityAvg = 0.8f * rescueState.sensor.zVelocityAvg + rescueState.sensor.zVelocity * 0.2f;
+        const timeUs_t currentTimeUs = micros();
+        const timeDelta_t gpsDataIntervalUs = cmpTimeUs(currentTimeUs, previousDataTimeUs);
+        rescueState.sensor.gpsDataIntervalSeconds = constrainf(gpsDataIntervalUs * 0.000001f, 0.01f, 1.0f);
+        // Range from 10ms (100hz) to 1000ms (1Hz). Intended to cover common GPS data rates and exclude unusual values.
+        previousDataTimeUs = currentTimeUs;
 
-        rescueState.sensor.accMagnitude = (float) sqrtf(sq(acc.accADC[Z]) + sq(acc.accADC[X]) + sq(acc.accADC[Y])) * acc.dev.acc_1G_rec;
-        rescueState.sensor.accMagnitudeAvg = (rescueState.sensor.accMagnitudeAvg * 0.8f) + (rescueState.sensor.accMagnitude * 0.2f);
+        rescueState.sensor.filterK = pt1FilterGain(0.8, 1.0f / rescueState.sensor.gpsDataIntervalSeconds); 
```

Comment:
```suggestion
        rescueState.sensor.filterK = pt1FilterGain(0.8, rescueState.sensor.gpsDataIntervalSeconds); 
```

Is `gpsDataIntervalSeconds` a time duration (duration between GPS samples)? Then we need to get rid of the inversion.

Reply:
Yes, it is the time duration in seconds

Reply:
much smoother now :-)